### PR TITLE
Loader reconciliation preparatory changes

### PR DESCRIPTION
**Pull request by dylandreimerink**
We are slowly moving towards a loader that is resilient and reconciles whenever device changes are detected. Currently we have datapath logic to detect device changes, this needs to be reported up all the way to the Daemon which then pushes Reinitilization requests all the way down to the loader. This PR contains a number of preparatory changes that will make it easier for a followup PR to make the datapath / loader reconcile without going through the daemon, based on the devices table.

This PR removes the `BaseProgramOwner` which was one big layer violation. All dependencies the loader needs are now passed via Hive. Only the actual Reinitialize calling still exists, which will be replaced by device based reconciliation in a followup.

### CODE REVIEW

File: pkg/datapath/types/orchistrator.go
Comment (joamaki) [inline pkg/datapath/types/orchistrator.go]: File name typo

File: pkg/datapath/types/loader.go
Comment (lmb) [inline pkg/datapath/types/loader.go:83]: Can you document why this is a separate type please? I imagine it has something to do with the fact that it's taken from all over the place. Future me would also very much be interested to know at which point we could get rid of this.

**Review by lmb — CHANGES_REQUESTED** [review]
Loader changes look good to me.

**Review by lmb — APPROVED** [review]

**Review by tklauser — APPROVED** [review]
Looks good for @cilium/ci-structure (re-adding review requests for the other code owners which I didn't cover).

File: pkg/datapath/types/loader.go
Comment (gentoo-root) [inline pkg/datapath/types/loader.go]: Nit: deamon -> daemon.

File: daemon/cmd/config.go
Comment (gentoo-root) [inline daemon/cmd/config.go:1]: Nit: typo in the commit message: "orchistrator".

File: pkg/proxy/cell.go
Comment (gentoo-root) [inline pkg/proxy/cell.go:70]: Nit: typo "Manger".

**Review by gentoo-root — APPROVED** [review]
Great cleanup of dependencies!

**Review by tommyp1ckles — APPROVED** [review]

**Review by jrajahalme — APPROVED** [review]
LGTM for the proxy files.

**Comment by ti-mo** [comment]
A few unresolved conversations that should be addressed before merging. Looks like `ci-e2e-upgrade` is not required at the moment, but there's one connection interruption during upgrade. Could it be related?

**Pull request merged**
